### PR TITLE
rationalise and unpin pillow

### PR DESCRIPTION
**Pull request by bjlittle**
This PR rationalises the contents of the `tephi` test images and unpins `pillow`.

It appears that we almost have **twice as many images** that we actually need. 

We've been using imagehashing in `tephi` for only 5 releases, namely v2.0.0 through to v2.0.4. During which time the `tephi.tests.results.imagerepo.json` has **not** changed.

Currently, there are 22 graphical tests, of which 3 tests result in legitimate duplicate images. Therefore we should only have 19 images in the repo. This is the result of this PR.

In addition, I've also unpinned `pillow`, as this is now the case in `tephi`. The change in `pillow` from 6.x to 9.x results in **only one** existing image with a new hash (that has a hamming distance of 4) i.e.,

```
existing e92596db92db6d249e9a3386c64969c7c331964f0c9c69233c646e19e5b3c786.png
new      e92596db92db6d249e9b3386c6596987c330964f0c9c69233c646e19e5b3c786
                            x      x   x    x
```
This has simply been added as a symbolic link.

### CODE REVIEW

File: .cirrus.yml [hunk 1]
```diff
@@ -19,25 +19,14 @@ env:
   CACHE_PERIOD: "2"
   # Increment the build number to force new conda cache upload.
   CONDA_CACHE_BUILD: "0"
-  # Increment the build number to force new nox cache upload.
-  NOX_CACHE_BUILD: "0"
-  # Increment the build number to force new pip cache upload.
-  PIP_CACHE_BUILD: "0"
```
Comment (bjlittle) [inline .cirrus.yml:25]: This isn't used, so purging it

File: .cirrus.yml [hunk 2]
```diff
@@ -19,25 +19,14 @@ env:
...
 #
 test_task:
   auto_cancellation: true
-  matrix:
-    env:
-      PY_VER: "3.6"
-    env:
-      PY_VER: "3.7"
-    env:
-      PY_VER: "3.8"
```
Comment (bjlittle) [inline .cirrus.yml:39]: There is no need for the `matrix`, so purging it

File: .cirrus.yml [hunk 3]
```diff
@@ -48,6 +37,7 @@ test_task:
       - wget --quiet https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O miniconda.sh
       - echo "${CIRRUS_OS} $(sha256sum miniconda.sh)"
       - echo "$(date +%Y).$(expr $(date +%U) / ${CACHE_PERIOD}):${CONDA_CACHE_BUILD}"
+      - echo "$(sha256sum requirements.txt)"
```
Comment (bjlittle) [inline .cirrus.yml:40]: Let's capture changes in requirements as part of the `fingerprint`

File: run_test.py [hunk 4]
```diff
@@ -9,19 +9,23 @@
...
         for fname in glob.glob("images/*.png"):
             phash = imagehash.phash(Image.open(fname), hash_size=_HASH_SIZE)
             fname_base = os.path.basename(fname)
-            fname_hash = os.path.splitext(fname_base)[0]
-            if str(phash) != fname_hash:
-                msg = 'Calculated phash {} does not match filename {!r}.'
-                exceptions.append(ValueError(msg.format(str(phash), fname_base)))
+            fname_hash = imagehash.hex_to_hash(os.path.splitext(fname_base)[0])
+            hamming = fname_hash - phash
+            if hamming > _HAMMING_DISTANCE:
```
Comment (bjlittle) [inline run_test.py:24]: We really should be using hamming distances here and not string comparisons of the hashes.

I believe this is a source of unnecessary image proliferation within the repo

File: run_test.py [hunk 5]
```diff
@@ -9,19 +9,23 @@
 
 
 _HASH_SIZE = 16
+# Ideally, a hamming distance of 2 is typically sufficient. However,
+# we're relaxing this to 4 to accommodate the migration in behaviour
+# of unpinning pillow<7.
+_HAMMING_DISTANCE = 4
```
Comment (bjlittle) [inline run_test.py:14]: This seems like a pragmatic position to take here.

File: .cirrus.yml (hunk 3 above)
Comment (wjbenfold) [reply .cirrus.yml:40]: Does `CIRRUS_OS` include the python version or should that be captured here too?

File: run_test.py (hunk 5 above)
Comment (wjbenfold) [inline run_test.py:14]: I thought you were setting this to 2 given that you've added the extra image to account for the unpinning?

File: run_test.py (hunk 4 above)
Comment (wjbenfold) [reply run_test.py:24]: I wouldn't be averse to the use of `pathlib` here - it's the future and more readable. Seems fine as-is though

**Review by wjbenfold — CHANGES_REQUESTED** [review]

File: run_test.py (hunk 5 above)
Comment (bjlittle) [reply run_test.py:14]: The hamming distance requires to be at least 4 here, as detailed in the PR description.

But we can now revert the hamming distance in `tephi` itself back from `4` to `2` :+1:

File: .cirrus.yml (hunk 3 above)
Comment (bjlittle) [reply .cirrus.yml:40]: The python version shouldn't matter, we're just rolling with the latest version.

If `imagehash` gives different results depending on the version of `python`, then we're in BIG trouble :rofl:

File: run_test.py (hunk 5 above)
Comment (wjbenfold) [reply run_test.py:14]: I'd confused myself with which repo we're in, I'm there now, this is all good

File: run_test.py (hunk 4 above)
Comment (bjlittle) [reply run_test.py:24]: Done

**Review by wjbenfold — APPROVED** [review]

**Pull request merged**